### PR TITLE
Updated polygon class with static method to perform intersection analysis on adjacent 2D polygons.  Updated unit tests as well

### PR DESCRIPTION
**Pull request by DatHydroGuy**
I have added a static method to the Polygon2D class to perform intersection analysis on adjacent polygons.

### CODE REVIEW

**Comment by chriswmackey** [comment]
Thank you for doing this @DatHydroGuy . This is definitely a useful contribution that will be able to count towards hacktoberfest. I also like the way that you have implemented it. The insertion of new points in reverse order is particularly clever. However, to be able to merge this in, there are four changes that need to be made:

1. We need a docstring on your staticmethod here so that we can remember what its purpose is.
2. It would be extremely helpful if this method could take a `tolerance` input, which is the minimum distance from an endpoint to an adjacent `LineSegment3D` that the two are considered intersected. For this, you can use the [closest_point2d_on_line2d](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/intersection2d.py#L157-L174) method instead of the intersect method you are using now (sorry for the lack of clarity in the original issue).
3. The way that you have written it now, I think you will get duplicate vertices for Polygons in the case that two adjacent polygons share a vertex (we only want the intersection point to be added when it occurs along the edge; not at the ends). A simple check to be sure that the intersection point is not within the `tolerance` distance of the vertex should be able to catch this.
4. If you are able to add a test that is just like the 3-polygon sample cited in [the issue that you are addressing](https://github.com/ladybug-tools/ladybug-geometry/issues/57), that will assuage any concern about it running correctly.

If you address those 4 things, we can merge it in but I will also note that, in order to close [the issue that you are addressing](https://github.com/ladybug-tools/ladybug-geometry/issues/57), we ultimately need a staticmethod that accepts a list of Polygon2D objects as its input and computes all intersections across the whole list (not just an intersection between two polygons). I think we can do this by keeping the hidden method that you have now and just writing a public method that loops through all of the polygons in the list and uses your 2-polygon intersection method to get new intersected polygons. I think we may also want to add a "bounding rectangle" check in this public method that runs first to see if two polygons have any potential of overlapping since the whole intersection process can be rather intense compared to simply checking the `min` and `max` of the polygons. As, I said, you don't have to address this as part of this PR and you clearly laid enough groundwork here to ultimately solve the issue. But I describe this so that you know what we ultimately need.

Thanks again!

**Comment by DatHydroGuy** [comment]
Hi @chriswmackey 
Thank you for the detailed feedback!  I'll start looking at the points you've raised and submit an update.

**Comment by chriswmackey** [comment]
Thanks @DatHydroGuy . Just let me know if you have any questions as you go through the feedback. I should be much more responsive here on out now that I am past my wedding, which was last week.

**Comment by DatHydroGuy** [comment]
@chriswmackey - congratulations on the wedding!!!  I hope you and your significant other had a great day!
I'm almost done with the updates (I think), so I'll put in the pull request once I've convinced myself that I've covered off the scenarios and edge cases.

**Comment by DatHydroGuy** [comment]
Hi @chriswmackey 
I have updated the PR with a static method which takes a list of polygons and performs the segment intersect test on them to a given tolerance.
One unexpected thing I encountered with introducing a tolerance value was with floating-point errors which were causing tolerance checks to fail.  I have had to counter this by implementing the private _is_close method  in the Polygon2D class, which is a direct copy of the PEP0485 method.  By default, this performs a comparison to an accuracy of 10e-9, but this is configurable.  I wasn't sure about the best way to implement it, or even if it's an acceptable solution for you, so I'm happy to discuss the matter further.
The unit tests I've added all capture issues I've encountered whilst implementing this update, so I hope they offer "real world" value.
As always, happy for feedback and to discuss any of the details further.

File: ladybug_geometry/geometry2d/polygon.py [hunk 1]
```diff
@@ -717,6 +717,94 @@ def _are_clockwise(vertices):
...
+        # Bounding rectangle check
+        polygon1_width = polygon1.max.x - polygon1.min.x
+        polygon1_height = polygon1.max.y - polygon1.min.y
+        polygon2_width = polygon2.max.x - polygon2.min.x
+        polygon2_height = polygon2.max.y - polygon2.min.y
+        if (abs(polygon1.min.x - polygon2.min.x) * 2 <= (polygon1_width + polygon2_width + tolerance)) and \
+                (abs(polygon1.min.y - polygon2.min.y) * 2 <= (polygon1_height + polygon2_height + tolerance)):
+            pass
+        else:
+            return polygon1, polygon2
```
Comment (chriswmackey) [inline ladybug_geometry/geometry2d/polygon.py]: This is acceptable but I think you can make this a little more pythonic and terse here by avoiding the `else` with the following:
```
 if not (abs(polygon1.min.x - polygon2.min.x) * 2 <= (polygon1_width + polygon2_width + tolerance)) and \
                (abs(polygon1.min.y - polygon2.min.y) * 2 <= (polygon1_height + polygon2_height + tolerance)):
    return polygon1, polygon2
```

File: ladybug_geometry/geometry2d/polygon.py [hunk 2]
```diff
@@ -717,6 +717,94 @@ def _are_clockwise(vertices):
...
+
+        for i1, seg1 in enumerate(polygon1.segments):
+            for i2, seg2 in enumerate(polygon2.segments):
+                x = closest_point2d_on_line2d(seg2.p1, seg1)
+                if all([p.distance_to_point(x) > tolerance for p in polygon1.vertices]) and\
```
Comment (chriswmackey) [inline ladybug_geometry/geometry2d/polygon.py]: I hate to be the PEP8 👮 but I think there should be a whitespace after `and`.

File: ladybug_geometry/geometry2d/polygon.py [hunk 3]
```diff
@@ -717,6 +717,94 @@ def _are_clockwise(vertices):
...
+                        Polygon2D._is_close(x.distance_to_point(seg2.p1), tolerance):
+                    polygon1_updates.append([i1, x])
+                y = closest_point2d_on_line2d(seg1.p1, seg2)
+                if all([p.distance_to_point(y) > tolerance for p in polygon2.vertices]) and\
```
Comment (chriswmackey) [inline ladybug_geometry/geometry2d/polygon.py]: Same think here.

File: ladybug_geometry/geometry2d/polygon.py (hunk 2 above)
Comment (mostaphaRoudsari) [reply ladybug_geometry/geometry2d/polygon.py]: and if I'm not mistaken you don't need the brackets. Using a generator should work just fine and will be more efficient:

```suggestion
                if all(p.distance_to_point(x) > tolerance for p in polygon1.vertices) and \
```

Comment (chriswmackey) [reply ladybug_geometry/geometry2d/polygon.py]: Ah, good to know @mostaphaRoudsari . I think I should change this in some of my code as well.

File: ladybug_geometry/geometry2d/polygon.py (hunk 1 above)
Comment (DatHydroGuy) [reply ladybug_geometry/geometry2d/polygon.py]: Agreed.  Definitely looks cleaner without that pass statement in there.

File: ladybug_geometry/geometry2d/polygon.py (hunk 2 above)
Comment (DatHydroGuy) [reply ladybug_geometry/geometry2d/polygon.py]: @chriswmackey I think you're correct on that whitespace (although PyCharm didn't flag it as a PEP8 violation, oddly).  I've added it in anyway for consistency.
@mostaphaRoudsari that's a good catch!  The lazy-evaluation of generators is definitely more efficient.  Changes made.

File: ladybug_geometry/geometry2d/polygon.py (hunk 3 above)
Comment (DatHydroGuy) [reply ladybug_geometry/geometry2d/polygon.py]: Whitespace added.

**Comment by DatHydroGuy** [comment]
Hi @chriswmackey 
I have implemented the review suggestions that you and @mostaphaRoudsari made - thank you both for the feedback.
I have added extra comments to explain the main functional areas of the code, and added some comments into the unit tests as well to explain the assert statements.  I hope this provides clearer intent to the tests.
Further comments / suggestions welcome if you have them.

**Comment by chriswmackey** [comment]
Thanks for making the changes, @DatHydroGuy .  Let's merge this in. Not only is this going to be very useful in the code base but this is also a shining example of a hactoberfest PR. Thank you for setting a good example for others!

**Comment by DatHydroGuy** [comment]
@chriswmackey thank YOU for allowing me to contribute to such a great project!  It's been a very positive experience for me, and I've really appreciated all the feedback and input.

**Comment by lauleh** [comment]
Good evening @DatHydroGuy,
My name is Lauleh and I am reaching out from Ladybug Tools.
I want to personally thank you for making contributions to our repositories during the Hacktoberfest 2019.
If you can kindly take a moment to e-mail me (lauleh@ladybug.tools) your mailing address, I would like to send you a Ladybug Tools sticker. Thank you so much!

Best,

-- 
Lauleh Aslani | Administrative Assistant 
Ladybug Tools
www.ladybug.tools | lauleh@ladybug.tools
